pkg/permissions/clusterrole.go

```diff
@@ -0,0 +1,114 @@
+package permissions
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/argoproj-labs/argocd-operator/pkg/argoutil"
+	"github.com/argoproj-labs/argocd-operator/pkg/mutation"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrlClient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type ClusterRoleRequest struct {
+	Name              string
+	InstanceName      string
+	InstanceNamespace string
+	Labels            map[string]string
+	Annotations       map[string]string
+	Component         string
+	Rules             []rbacv1.PolicyRule
+
+	// array of functions to mutate role before returning to requester
+	Mutations []mutation.MutateFunc
+	Client    ctrlClient.Client
+}
+
+// newClusterRole returns a new clusterRole instance.
+func newClusterRole(name, instanceName, instanceNamespace, component string, labels, annotations map[string]string, rules []rbacv1.PolicyRule) *rbacv1.ClusterRole {
+	crName := argoutil.GenerateUniqueResourceName(instanceName, instanceNamespace, component)
+	if name != "" {
+		crName = name
+	}
+
+	return &rbacv1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        crName,
+			Labels:      argoutil.MergeMaps(argoutil.LabelsForCluster(instanceName, component), labels),
+			Annotations: argoutil.MergeMaps(argoutil.AnnotationsForCluster(instanceName, instanceNamespace), annotations),
+		},
+		Rules: rules,
+	}
+}
+
+func RequestClusterRole(request ClusterRoleRequest) (*rbacv1.ClusterRole, error) {
```

Comment:
Add description to the functions that do not have descriptions

Reply:
done